Workflow/Tests/WorkflowObserverTests.swift

```diff
@@ -0,0 +1,537 @@
+            willRenderSessions.map(\.workflowTypeString),
+            ["\(Parent.self)", "\(Child.self)"]
+        )
+
+        XCTAssertEqual(
+            didRenderSessions.map(\.workflowTypeString),
+            ["\(Child.self)", "\(Parent.self)"]
+        )
+    }
+
+    func test_didChangeCallbacks() {
+        var didChangeCallCount = 0
+        observer.onDidChange = { old, new, state, session in
+            guard old is Child, new is Child else {
+                XCTFail("unexpected workflow type")
+                return
+            }
+
+            didChangeCallCount += 1
+        }
+
+        let node = WorkflowNode(
+            workflow: Parent(),
+            parentSession: nil,
+            observer: observer
+        )
+
+        _ = node.render()
+
+        XCTAssertEqual(didChangeCallCount, 0)
+
+        _ = node.render()
+
+        XCTAssertEqual(didChangeCallCount, 1)
+    }
+
+    func test_didReceiveActionCallbacks() {
+        var actions: [StateTransitioningWorkflow.Event] = []
+        observer.onDidReceiveAction = { action, workflow, session in
+            guard let action = action as? StateTransitioningWorkflow.Event else {
+                XCTFail("unexpected action")
+                return
+            }
+
+            actions.append(action)
+        }
+
+        let node = WorkflowNode(
+            workflow: StateTransitioningWorkflow(),
+            parentSession: nil,
+            observer: observer
+        )
+
+        let rendering = node.render()
+        node.enableEvents()
+
+        XCTAssertEqual(actions, [])
+
+        rendering.toggle()
+
+        XCTAssertEqual(actions, [.toggle])
+    }
+
+    func test_willApplyActionCallbacks() {
+        var willApplyActions: [StateTransitioningWorkflow.Event] = []
+        var didApplyActions: [StateTransitioningWorkflow.Event] = []
+        var willApplyState: Bool?
+        var didApplyState: Bool?
+
+        observer.onApplyAction = { action, workflow, state, session in
+            let anyAction = action as? AnyWorkflowAction<StateTransitioningWorkflow>
+            guard let action: StateTransitioningWorkflow.Event = anyAction?.getWrappedValue() else {
+                XCTFail("unexpected action type")
+                return nil
+            }
+
+            willApplyState = state as? Bool
+            willApplyActions.append(action)
+
+            return { state, output in
+                didApplyState = state as? Bool
+                didApplyActions.append(action)
+            }
+        }
+
+        let node = WorkflowNode(
+            workflow: StateTransitioningWorkflow(),
+            parentSession: nil,
+            observer: observer
+        )
+
+        let rendering = node.render()
+        node.enableEvents()
+
+        XCTAssertEqual(willApplyActions, [])
+        XCTAssertEqual(didApplyActions, [])
+
+        rendering.toggle()
+
+        XCTAssertEqual(willApplyState, false)
+        XCTAssertEqual(didApplyState, true)
```

Comment:
`XCTAssertTrue/False`

Reply:
the values are optional, so IMO this reads better

Reply:
Ah! I missed that they were optional